tests/quality_pragmaitc/test_refinement.py

```diff
@@ -24,63 +30,112 @@ def test_shouldBeApplicable():
 
-        #assert False == goal.isApplicable(fullContext)
-        assert False == task.isApplicable(fullContext)
-        assert False == delegation.isApplicable(fullContext)
+    task.addApplicableContext(currentContext)
+    task.setProvidedQuality(currentContext, CommonMetrics.SECONDS, 12)
 
-    def test_aTaskShouldBeAchievable():
-        commonMetrics = CommonMetrics()
-        task = Task()
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
 
-        currentContext = Context("C1")
-        fullContext = []
-        fullContext.append(currentContext)
+    assert task in task.isAchievable(
+        fullContext, interp).getTasks()
 
-        qc = QualityConstraint(currentContext, commonMetrics.SECONDS, 15, 'LESS_OR_EQUAL_TO')
 
-        task.addApplicableContext(currentContext)
-        task.setProvidedQuality(currentContext, commonMetrics.SECONDS, 12)
+def test_shouldAddSeveralContextsAtOnce():
+    context1 = Context("C1")
+    context2 = Context("C2")
 
-        interp = Interpretation()
-        interp.addQualityConstraint(qc)
+    task = Task()
+    originalSize = 0
 
-        assert True == task in task.isAchievable(fullContext, interp).getTasks()
+    if None is task.getApplicableContext():
+        originalSize = len(task.getApplicableContext())
 
-    def test_shouldAddSeveralContextsAtOnce():
-        context1 = Context("C1")
-        context2 = Context("C2")
+    set = []
+
+    set.append(context1)
+    set.append(context2)
+
+    task.addApplicableContext(set)
+
+    assert 2 == len(task.getApplicableContext()) - originalSize
+
+
+def aNonApplicableRootGoalIsNotAchievable():
+    goal = Goal(False)
+    current = Context("C1")
+    fullContext = []
+
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    goal.addApplicableContext(Context("C2"))
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    assert goal.isAchievable(fullContext, interp) is None
+
+
+def aGoalWithATaskMayBeAchievable():
+    goal = Goal(False)
+
+    task = Task()
+
+    current = Context("C1")
+    fullContext = []
+    fullContext.append(current)
 
-        task = Task()
-        originalSize = task.getApplicableContext().len()
-        set = []
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
 
-        set.append(context1)
-        set.append(context2)
+    task.addApplicableContext(current)
+    task.setProvidedQuality(current, CommonMetrics.SECONDS, 13)
 
-        task.addApplicableContext(set)
+    goal.addDependency(task)
+    #goal.setIdentifier("Root")
+    goal.addApplicableContext(current)
 
-        #null is always an applicable context
-        assert 2 == task.getApplicableContext().len() - originalSize
\ No newline at end of file
+    plan = goal.isAchievable(fullContext, interp)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=4861978)